keeper/query_server_v2.go

```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: BUSL-1.1
+//
+// Copyright (C) 2025, NASD Inc. All rights reserved.
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file of this repository and at www.mariadb.com/bsl11.
+//
+// ANY USE OF THE LICENSED WORK IN VIOLATION OF THIS LICENSE WILL AUTOMATICALLY
+// TERMINATE YOUR RIGHTS UNDER THIS LICENSE FOR THE CURRENT AND ALL OTHER
+// VERSIONS OF THE LICENSED WORK.
+//
+// THIS LICENSE DOES NOT GRANT YOU ANY RIGHT IN ANY TRADEMARK OR LOGO OF
+// LICENSOR OR ITS AFFILIATES (PROVIDED THAT YOU MAY USE A TRADEMARK OR LOGO OF
+// LICENSOR AS EXPRESSLY REQUIRED BY THIS LICENSE).
+//
+// TO THE EXTENT PERMITTED BY APPLICABLE LAW, THE LICENSED WORK IS PROVIDED ON
+// AN "AS IS" BASIS. LICENSOR HEREBY DISCLAIMS ALL WARRANTIES AND CONDITIONS,
+// EXPRESS OR IMPLIED, INCLUDING (WITHOUT LIMITATION) WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT, AND
+// TITLE.
+
+package keeper
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"cosmossdk.io/collections"
+	"cosmossdk.io/errors"
+	"cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	transfertypes "github.com/cosmos/ibc-go/v8/modules/apps/transfer/types"
+	solomachine "github.com/cosmos/ibc-go/v8/modules/light-clients/06-solomachine"
+	tendermint "github.com/cosmos/ibc-go/v8/modules/light-clients/07-tendermint"
+
+	"dollar.noble.xyz/v2/types"
+	"dollar.noble.xyz/v2/types/v2"
+)
+
+var _ v2.QueryServer = &queryServerV2{}
+
+type queryServerV2 struct {
+	*Keeper
+}
+
+func NewQueryServerV2(keeper *Keeper) v2.QueryServer {
+	return &queryServerV2{Keeper: keeper}
+}
+
+func (k queryServerV2) Stats(ctx context.Context, req *v2.QueryStats) (*v2.QueryStatsResponse, error) {
+	if req == nil {
+		return nil, types.ErrInvalidRequest
+	}
+
+	stats, err := k.Keeper.Stats.Get(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get stats from state")
+	}
+
+	totalExternalYield := make(map[string]v2.QueryStatsResponse_ExternalYield)
+	for key, rawAmount := range stats.TotalExternalYield {
+		splitKey := strings.Split(key, "/")
+		provider := v2.Provider(v2.Provider_value[splitKey[0]])
+		identifier := splitKey[1]
```

Comment:
this is at least the second time i have seen this logic which could be an indicator that it should be broken out into a separate function.